flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/clustering/KMeans.scala

```diff
@@ -0,0 +1,614 @@
+    parameters.add(NumIterations, numIterations)
+    this
+  }
+
+  /**
+   * Sets the number of clusters.
+   *
+   * @param numClusters The number of clusters
+   * @return itself
+   */
+  def setNumClusters(numClusters: Int): KMeans = {
+    parameters.add(NumClusters, numClusters)
+    this
+  }
+
+  /**
+   * Sets the initial centroids on which the algorithm will start computing. These points should
+   * depend on the data and will significantly influence the resulting centroids.
+   * Note that this setting will override [[setInitializationStrategy())]] and the size of
+   * initialCentroids will override the value, if set, by [[setNumClusters()]]
+   *
+   * @param initialCentroids A set of labeled vectors.
+   * @return itself
+   */
+  def setInitialCentroids(initialCentroids: Seq[LabeledVector]): KMeans = {
+    parameters.add(InitialCentroids, initialCentroids)
+    this
+  }
+
+  /**
+   * Automatically initialize the KMeans algorithm. Allowed options are "random", "kmeans++" and
+   * "kmeans||"
+   *
+   * @param initialStrategy
+   * @return itself
+   */
+  def setInitializationStrategy(initialStrategy: String): KMeans = {
+    require(Array("random", "kmeans++", "kmeans||").contains(initialStrategy), s"$initialStrategy" +
+      s" is not supported")
+    parameters.add(InitialStrategy, initialStrategy)
+    this
+  }
+
+  /**
+   * Oversampling factor to be used in case the initialization strategy is set to be "kmeans||"
+   *
+   * @param oversamplingFactor Oversampling factor(\ell)
+   * @return this
+   */
+  def setOversamplingFactor(oversamplingFactor: Double): KMeans = {
+    require(oversamplingFactor > 0, "Oversampling factor must be positive.")
+    parameters.add(OversamplingFactor, oversamplingFactor)
+    this
+  }
+
+  /**
+   * Number of initialization rounds to be done when the initialization strategy is set to be
+   * "kmeans||"
+   *
+   * @param numRounds Number of rounds(r)
+   * @return this
+   */
+  def setNumRounds(numRounds: Int): KMeans = {
+    require(numRounds > 0, "Number of rounds must be positive")
+    parameters.add(KMeansParRounds, numRounds)
+    this
+  }
+
+}
+
+/**
+ * Companion object of KMeans. Contains convenience functions, the parameter type definitions
+ * of the algorithm and the [[FitOperation]] & [[PredictOperation]].
+ */
+object KMeans {
+
+  private val RANDOM_FRACTION = "random_sample_fraction"
+  private val PARINIT_SET = "par_init_solution_set"
+  private val PARINIT_COST = "par_init_solution_cost"
+  private val PARINIT_SAMPLE = "par_init_oversample_factor"
+
+  /** Euclidean Distance Metric */
+  val euclidean = EuclideanDistanceMetric()
+
+  case object NumIterations extends Parameter[Int] {
+    val defaultValue = Some(10)
+  }
+
+  case object InitialCentroids extends Parameter[Seq[LabeledVector]] {
+    val defaultValue = None
+  }
+
+  case object InitialStrategy extends Parameter[String]{
+    val defaultValue = Some("kmeans||")
```

Comment:
"kmeans||" seems awkward I would expect a something like KMeansParallel... at least in the code if not in the description...